File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque

class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        
        # dfs to create a associated list

        parent = defaultdict(TreeNode)

        def dfs(node):
            if node.left:
                parent[node.left] = node
                dfs(node.left)
            if node.right:
                parent[node.right] = node
                dfs(node.right)
            return

        dfs(root)
        bfs_q = deque()
        visited = set()
        # | | | | |
        # append(): Insert from rear/right
        # popleft(): delete from front/left
        

        bfs_q.append((target,0))
        res = []
        

        node_dist_dict = defaultdict(int)

        while len(bfs_q)>0:
            node,height = bfs_q.popleft()

            visited.add(node)
            
            # node_dist_dict[node]=height
            if height == k:
                res.append(node.val)
            
            if height > k:
                break

            
            if node in parent.keys() and parent[node] not in visited  :
                bfs_q.append((parent[node],height+1))

            if node.left and node.left not in visited :
                bfs_q.append((node.left,height+1))

            if node.right and node.right not in visited:
                bfs_q.append((node.right,height+1))

        

        return res  
        


        




        

File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */


class Solution {
public:

    void dfs(TreeNode *node, std::unordered_map<TreeNode *, TreeNode *> *parentMap)
    {
        // Insert logic here
        // left traversal
        if (node->left != NULL)
        {
            /*TreeNode* is a pointer and it is a valid key type for std::unordered_map, because pointers can be hashed and compared by default.
            The actual issue seems to be related to how you're accessing the parentMap.
            Since parentMap is a pointer to an unordered_map, you need to dereference it before using the subscript operator ([]).
            You should be using (*parentMap)[child] = node; instead of parentMap[child] = node;.*/
            (*parentMap)[node->left] = node;
            dfs(node->left, parentMap);
        }
        // right traversal
        // right traversal
        if (node->right != NULL)
        {
            (*parentMap)[node->right] = node;
            dfs(node->right, parentMap);
        }
    }
    
    

    vector<int> distanceK(TreeNode *root, TreeNode *target, int k)
    {
        // create a parent mapping
        // do it with recursive dfs
        std::unordered_map<TreeNode *, TreeNode *> parentMap;
        // BFS Q
        queue<pair<TreeNode *, int>> bfsQ;
        // visited array
        vector<bool>visited(500);
        // result
        vector<int> res;


        // populate parents
        dfs(root, &parentMap);
        // init BFS
        bfsQ.push({target, 0});
        // Exec BFS 
        while (!bfsQ.empty())
        {

            pair<TreeNode *, int> package = bfsQ.front();
            bfsQ.pop();

            TreeNode *node = package.first;

            if ( node == NULL ) {
                continue;
            }

            if (visited[node->val]){
                
                // cout<< "revisiting" << node->val <<endl;
                continue;
            }

            int dist = package.second;
            visited[node->val] = true;
            
            // cout << "-----at" <<(*node).val <<endl;

            if (dist == k)
            {
                // stop further exploration
                
                // cout << (*node).val << "dist from " << (*target).val << "is " << dist<<endl;;
                res.push_back(node->val);
                continue;
            }
            // explore LEFT
            if ((node->left != NULL) )
            {   
                
                // cout<<"inserting left" << node->left->val << endl;
                bfsQ.push({node->left, dist + 1});
            }
            // explore RIGHT
            if ((node->right != NULL) )
            {   
                
                // cout<<"inserting right" << node->right->val << endl;
                bfsQ.push({node->right, dist + 1});
            }
            // explore Parent-wise
            
            // cout << "(parentMap[node] != NULL)" <<(parentMap[node] != NULL)   ;

            if ((parentMap[node] != NULL) )
            {   
                // cout<<"inserting parent" << parentMap[node]->val << endl;
                bfsQ.push({parentMap[node], dist + 1});
            }
        }

        return res;
    }
};